src/tracing/esp32_diagnostic_trace/Diagnostics.h

```diff
@@ -0,0 +1,249 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/TLVCircularBuffer.h>
+#define kMaxStringValueSize 128
+
+namespace chip {
+namespace Tracing {
+
+namespace Diagnostics {
+
+// Diagnostic TAGs
+enum class DiagTag : uint8_t
+{
+    TIMESTAMP = 0,
+    LABEL,
+    VALUE,
+};
+
+/**
+ * @class DiagnosticEntry
+ * @brief Abstract base class for encoding diagnostic entries into TLV format.
+ *
+ */
+class DiagnosticEntry
+{
+public:
+    /**
+     * @brief Virtual destructor for proper cleanup in derived classes.
+     */
+    virtual ~DiagnosticEntry() = default;
+
+    /**
+     * @brief Pure virtual method to encode diagnostic data into a TLV structure.
+     *
+     * @param writer A reference to the `chip::TLV::CircularTLVWriter` instance
+     *               used to encode the TLV data.
+     * @return CHIP_ERROR Returns an error code indicating the success or
+     *                    failure of the encoding operation.
+     */
+    virtual CHIP_ERROR Encode(chip::TLV::CircularTLVWriter & writer) const = 0;
+
+    virtual CHIP_ERROR Decode(chip::TLV::TLVReader & reader) = 0;
+};
+
+template <typename T>
+class Diagnostic : public DiagnosticEntry
+{
+public:
+    Diagnostic(char * label = nullptr, T value = T{}, uint32_t timestamp = 0) : label_(label), value_(value), timestamp_(timestamp)
+    {}
+
+    CHIP_ERROR Encode(chip::TLV::CircularTLVWriter & writer) const override
+    {
+        chip::TLV::TLVType DiagnosticOuterContainer = chip::TLV::kTLVType_NotSpecified;
+        ReturnErrorOnFailure(
+            writer.StartContainer(chip::TLV::AnonymousTag(), chip::TLV::kTLVType_Structure, DiagnosticOuterContainer));
+
+        // Write timestamp
+        ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(DiagTag::TIMESTAMP), timestamp_));
+
+        // Write label
+        if (strlen(label_) > kMaxStringValueSize)
+        {
+            char labelBuffer[kMaxStringValueSize + 1];
+            memcpy(labelBuffer, label_, kMaxStringValueSize);
+            labelBuffer[kMaxStringValueSize] = '\0';
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), labelBuffer));
+        }
+        else
+        {
+            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), label_));
```

Comment:
may be we can use `PutString()`'s this variant which takes length as input, and we don't need an additional buffer, and length is enough? https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/TLVWriter.h#L585

```suggestion
            ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::LABEL), label_), std::min(strlen(label_), kMaxStringValueSize));
```